### PR TITLE
Add cache invalidation by age

### PR DESCRIPTION
**Pull request by stefanv**
This is a POC to illustrate cache invalidation by age.

Note that, currently, cache invalidation happens only on insert.  We can modify this if it seems more prudent to do it on get, or even in both places.

### CODE REVIEW

**Review by acrellin — CHANGES_REQUESTED** [review]
Can you also change `fn` to `filename` throughout `cache`? It took me a while to figure out what that variable was supposed to refer to because of the name.

File: skyportal/utils/cache.py [hunk 1]
```diff
@@ -71,20 +76,40 @@ def __setitem__(self, name, data):
...
+        cached_files = [
+            (f.stat().st_mtime, f.absolute()) for f in self._cache_dir.glob('*')
+        ]
+        cached_files = sorted(cached_files, key=lambda x: x[0], reverse=True)
+
+        now = time.time()
+
+        if self._max_age is not None:
+            removed_by_time = [
+                f for (t, f) in cached_files if (now - t) > self._max_age
```
Comment (acrellin) [inline skyportal/utils/cache.py:107]:
```suggestion
                filepath for (mod_time, fiilepath) in cached_files if (now - mod_time) > self._max_age
```
for readability

Comment (acrellin) [reply skyportal/utils/cache.py:107]: and wherever else applicable

Comment (stefanv) [reply skyportal/utils/cache.py:107]: Sure, I'll push that up.

File: .flake8 [hunk 2]
```diff
@@ -13,4 +13,4 @@
 [flake8]
 max-line-length = 80
 select = C,E,F,W,B,B950
-ignore = E501, W503
+ignore = E501, W503, E203
```
Comment (acrellin) [inline .flake8:16]: I don't think we _always_ want to ignore this rule, though, right?

File: skyportal/tests/utils/test_cache.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ def cache_parent_dir(tmpdir_factory):
 @pytest.fixture()
 def cache(cache_parent_dir):
     cache_path = pjoin(cache_parent_dir, 'cache')
-    cache = Cache(cache_path, max_items=3)
+    cache = Cache(cache_path, max_items=3, max_age=3)
```
Comment (acrellin) [inline skyportal/tests/utils/test_cache.py:21]: Please also make max_items nullable when using max_age

File: skyportal/utils/cache.py [hunk 4]
```diff
@@ -63,28 +68,50 @@ def __setitem__(self, name, data):
...
+
+        if self._max_age is not None:
+            removed_by_time = [
+                filename
+                for (mtime, filename) in cached_files
+                if (now - mtime) > self._max_age
+            ]
+            self._remove(removed_by_time)
+
+        oldest = cached_files[self._max_items :]
```
Comment (acrellin) [inline skyportal/utils/cache.py:113]: Please update to make `max_items` optional -- I want to only specify `max_age` w/ no limit on # cache items.

**Comment by acrellin** [comment]
@stefanv also, is there a way around the asymmetrical save/load mechanisms? It's counter-intuitive that `cache` creates the file for you when setting the cache but then just returns the filename rather than the corresponding Python object when loading the cache contents.

File: .flake8 (hunk 2 above)
Comment (stefanv) [reply .flake8:16]: I think we do; that is what is suggested by the Black documentation.

Comment (stefanv) [reply .flake8:16]: https://github.com/psf/black/issues/315#issuecomment-395457972

**Comment by stefanv** [comment]
@acrellin We can certainly modify this in any way that makes sense.

Re: creation: you have to generate the filename and create the file at some point: upon cache insertion seems like the right time.  If you want to change the API, I suggest we return the file object from which the filename & bytes can be accessed (`f.name`, `f.read()`).  I think the existing use cases demanded a filename, hence that is what I used.

The changed API would save having to do `with open(x, 'rb')`.

File: skyportal/utils/cache.py (hunk 4 above)
Comment (stefanv) [reply skyportal/utils/cache.py]: Done

**Comment by stefanv** [comment]
Failures unrelated.

**Review by acrellin — APPROVED** [review]

**Pull request merged**